File: windows-native/DataMover.Core/Services/IOSettings.cs
using System.Diagnostics;

namespace DataMover.Core.Services;

/// <summary>
/// Setari configurabile de I/O si memorie - port 1:1 al IOSettings.swift
/// (Mac) / core/io_settings.py (Windows Python). Persistat prin
/// AppSettingsStore (JSON in %AppData%), la fel ca restul setarilor.
/// </summary>
public sealed class IOPerformancePreset
{
    public string Id { get; init; } = "";
    public string Label { get; init; } = "";
    public int ChunkSizeMB { get; init; }
    public int RamLimitMB { get; init; }
}

public static class IOSettings
{
    public const int DefaultChunkSizeMB = 8;

    // Trepte granulate (cerinta explicita) - identice cu Mac/Python.
    public static readonly int[] ChunkSizeChoicesMB = { 1, 2, 4, 8, 16, 32, 64, 128 };
    public static readonly int[] RamLimitChoicesMB = { 0, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536 };

    public static readonly IOPerformancePreset[] Presets =
    {
        new() { Id = "eco", Label = "Eco / Sistem Slab", ChunkSizeMB = 4, RamLimitMB = 1024 },
        new() { Id = "standard", Label = "Standard", ChunkSizeMB = 8, RamLimitMB = 4096 },
        new() { Id = "high", Label = "Performanta Inalta", ChunkSizeMB = 32, RamLimitMB = 16384 },
        new() { Id = "extreme", Label = "Extrem / Productie RAW", ChunkSizeMB = 64, RamLimitMB = 32768 },
    };

    /// "512 MB" sub 1 GB, "8 GB" de la 1024 MB in sus.
    public static string SizeLabel(int mb) => mb >= 1024 ? $"{mb / 1024} GB" : $"{mb} MB";

    /// Memoria de lucru (Working Set) a procesului curent, in bytes -
    /// echivalentul IOSettings.currentResidentMemoryBytes() de pe Mac.
    public static long CurrentProcessMemoryBytes()
    {
        try
        {
            using var proc = Process.GetCurrentProcess();
            return proc.WorkingSet64;
        }
        catch
        {
            return 0;
        }
    }

    /// Backpressure: daca procesul a depasit ramLimitMb configurat,
    /// asteapta putin (verificand cancel-ul intre timpi) inainte sa lase
    /// urmatorul fisier sa inceapa - identic cu waitIfOverRAMLimit (Mac) /
    /// wait_if_over_ram_limit (Python).
    public static void WaitIfOverRamLimit(int ramLimitMb, CancelToken cancel, Action<string>? onWarning = null)
    {
        if (ramLimitMb <= 0) return;
        long limitBytes = (long)ramLimitMb * 1024 * 1024;
        long used = CurrentProcessMemoryBytes();
        if (used <= limitBytes) return;

        double waited = 0;
        bool warned = false;
        while (used > limitBytes && waited < 30.0)
        {
            if (cancel.IsCancelled) return;
            if (!warned)
            {
                onWarning?.Invoke(
                    $"ATENTIE: memoria aplicatiei ({used / 1024 / 1024} MB) a depasit limita " +
                    $"setata ({ramLimitMb} MB) - se asteapta putin inainte de urmatorul fisier (backpressure).");
                warned = true;
            }
            Thread.Sleep(500);
            waited += 0.5;
            used = CurrentProcessMemoryBytes();
        }
    }
}

File: mac-native/Sources/DataMoverMac/IOSettings.swift
import Foundation

/// Setari configurabile de I/O si memorie (2026-08-27) - port Swift al
/// core/io_settings.py. Motiv: raportat un caz real de "Your system has
/// run out of application memory" / swap la maxim, la un transfer de 3 TB
/// - vezi fix-ul de autoreleasepool din copyFileCancelable/genericHash
/// (OffloadEngine.swift), plus aceste setari configurabile de buffer si
/// plafon de memorie (Regula globala "Memory & I/O Performance" din
/// CLAUDE.md).
/// Preset rapid de performanta (2026-08-28) - combina o treapta de buffer
/// cu un plafon de RAM, ca userul sa nu trebuiasca sa aleaga doua valori
/// separat de fiecare data. Valorile individuale (chunkSizeMB/ramLimitMB)
/// raman oricand ajustabile manual dupa aplicarea unui preset.
struct IOPerformancePreset: Identifiable {
    let id: String
    let nameKey: String
    let chunkSizeMB: Int
    let ramLimitMB: Int

    static let all: [IOPerformancePreset] = [
        IOPerformancePreset(id: "eco", nameKey: "io.preset.eco", chunkSizeMB: 4, ramLimitMB: 1024),
        IOPerformancePreset(id: "standard", nameKey: "io.preset.standard", chunkSizeMB: 8, ramLimitMB: 4096),
        IOPerformancePreset(id: "high", nameKey: "io.preset.high", chunkSizeMB: 32, ramLimitMB: 16384),
        IOPerformancePreset(id: "extreme", nameKey: "io.preset.extreme", chunkSizeMB: 64, ramLimitMB: 32768),
    ]
}

enum IOSettings {
    // Trepte granulate (2026-08-28, cerinta explicita) - de la masini
    // modeste (carduri SD pe un Mac mini de baza) pana la statii de
    // productie RAW cu zeci de GB de RAM disponibili.
    static let chunkSizeChoicesMB = [1, 2, 4, 8, 16, 32, 64, 128]
    static let ramLimitChoicesMB = [0, 512, 1024, 2048, 4096, 8192, 16384, 32768, 65536] // 0 = fara limita
    static let defaultChunkSizeMB = 8

    private static let chunkKey = "datamover_chunk_size_mb"
    private static let ramLimitKey = "datamover_ram_limit_mb"

    static var chunkSizeMB: Int {
        get {
            let saved = UserDefaults.standard.integer(forKey: chunkKey)
            return saved > 0 ? saved : defaultChunkSizeMB
        }
        set { UserDefaults.standard.set(newValue, forKey: chunkKey) }
    }

    static var chunkSizeBytes: Int { chunkSizeMB * 1024 * 1024 }

    /// 0 = fara limita configurata de user.
    static var ramLimitMB: Int {
        get { UserDefaults.standard.object(forKey: ramLimitKey) as? Int ?? 1024 }
        set { UserDefaults.standard.set(newValue, forKey: ramLimitKey) }
    }

    static var ramLimitBytes: UInt64 { ramLimitMB > 0 ? UInt64(ramLimitMB) * 1024 * 1024 : 0 }

    /// Memoria rezidenta (RSS) a procesului curent, in bytes - foloseste
    /// task_info/mach_task_basic_info, singura API stabila pentru asta pe
    /// macOS (Foundation/AppKit nu expun un echivalent direct).
    static func currentResidentMemoryBytes() -> UInt64? {
        var info = mach_task_basic_info()
        var count = mach_msg_type_number_t(MemoryLayout<mach_task_basic_info>.size / MemoryLayout<natural_t>.size)
        let result: kern_return_t = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: Int(count)) {
                task_info(mach_task_self_, task_flavor_t(MACH_TASK_BASIC_INFO), $0, &count)
            }
        }
        guard result == KERN_SUCCESS else { return nil }
        return info.resident_size
    }

    /// Backpressure simplu: daca procesul a depasit ramLimitMB configurat,
    /// asteapta putin (verificand cancelul intre timpi) inainte sa lase
    /// urmatorul fisier sa inceapa - previne acumularea nestapanita in
    /// RAM/swap cand scrierea (ex. HDD) e mai lenta decat citirea (ex.
    /// SSD). E o limita ORIENTATIVA la nivel de proces, nu un plafon dur
    /// impus de OS - scopul e sa incetineasca sursa, nu sa garanteze un
    /// maxim absolut.
    static func waitIfOverRAMLimit(cancel: CancelToken, onWarning: ((String) -> Void)? = nil) {
        let limit = ramLimitBytes
        guard limit > 0 else { return }
        guard var used = currentResidentMemoryBytes(), used > limit else { return }

        var waited = 0.0
        var warned = false
        while used > limit && waited < 30.0 {
            if cancel.isCancelled { return }
            if !warned {
                onWarning?(
                    "ATENTIE: memoria aplicatiei (\(used / 1024 / 1024) MB) a depasit limita "
                    + "setata (\(limit / 1024 / 1024) MB) - se asteapta putin inainte de "
                    + "urmatorul fisier (backpressure)."
                )
                warned = true
            }
            Thread.sleep(forTimeInterval: 0.5)
            waited += 0.5
            guard let next = currentResidentMemoryBytes() else { break }
            used = next
        }
    }
}
